src/main/java/net/neoforged/neoforge/codec/IndexedNonNullListCodec.java

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.codec;
+
+import com.mojang.datafixers.util.Pair;
+import com.mojang.serialization.Codec;
+import com.mojang.serialization.DataResult;
+import com.mojang.serialization.DynamicOps;
+import com.mojang.serialization.MapCodec;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.function.Predicate;
+import net.minecraft.core.NonNullList;
+
+public record IndexedNonNullListCodec<E>(MapCodec<E> elementCodec, E defaultValue, Predicate<E> skipPredicate, String indexKey) implements Codec<NonNullList<E>> {
+    public IndexedNonNullListCodec(MapCodec<E> elementCodec, E defaultValue, Predicate<E> skipPredicate) {
+        this(elementCodec, defaultValue, skipPredicate, "index");
+    }
+
+    @Override
+    public <T> DataResult<Pair<NonNullList<E>, T>> decode(DynamicOps<T> ops, T input) {
+        var asMap = ops.getMap(input);
+        if (asMap.isError()) {
+            String err = asMap.error().map(DataResult.Error::message).orElse("Not a map.");
+            return DataResult.error(() -> err, Pair.of(NonNullList.create(), input));
+        }
+
+        return asMap.map(map -> {
+            int maxSize = Codec.INT.fieldOf("size")
+                    .decode(ops, map)
+                    .getOrThrow();
+
+            var finalList = NonNullList.<E>withSize(maxSize, defaultValue);
+            final var decoder = new NonNullElementDecoder<T, E>(ops, elementCodec, indexKey, finalList::set, (el) -> {});
+
+            ops.getList(map.get("items")).ifSuccess(stream -> {
+                stream.accept(decoder::process);
+            });
```

Comment:
I think the failure should be reported here rather than consumed into the ether.